dali/python/nvidia/dali/_debug_mode.py

```diff
@@ -114,6 +118,110 @@ def __rxor__(self, other):
+
+    def __init__(
+            self, source=None, num_outputs=None, batch_size=-1, cycle=None, name=None, layout=None,
+            batch=None, batch_info=None):
+        if name is not None and num_outputs is not None:
+            raise ValueError("`num_outputs` is not compatible with named `ExternalSource`")
+
+        callback, source_desc = _get_callback_from_source(source, cycle, batch_info or False)
+
+        self._name = name
+        self._layout = layout
+        self._num_outputs = num_outputs
+        self._batch = batch
+        self._batch_size = batch_size
+        self._callback = callback
+        self._source_desc = source_desc
+        self._batch_info = batch_info
+        self._current_iter = 0
+        self._current_sample = 0
+        self._feed_inputs = Queue()
+
+        if callback is not None:
+            arg_count = _accepted_arg_count(callback)
+            if arg_count not in [0, 1]:
+                raise TypeError("External source callback must be a callable with 0 or 1 argument")
+            self.accepts_arg = arg_count > 0
+
+    def _callback_args(self, idx_in_batch, epoch_idx):
+        if not self.accepts_arg:
+            return ()
+        if idx_in_batch is not None:
+            arg = _types.SampleInfo(
+                self._current_sample + idx_in_batch,
+                idx_in_batch,
+                self._current_iter,
+                epoch_idx)
+        elif self._batch_info:
+            arg = _types.BatchInfo(
+                self._current_iter,
+                epoch_idx)
+        else:
+            arg = self._current_iter
+        return (arg,)
+
+    def _get_batch(self, epoch_idx):
+        try:
+            if self._batch:
+                callback_out = self._callback(*self._callback_args(None, epoch_idx))
+            else:
+                callback_out = [self._callback(*self._callback_args(i, epoch_idx))
+                                for i in range(self._batch_size)]
+            self._current_sample += self._batch_size
+            self._current_iter += 1
+        except StopIteration:
+            self._current_iter = 0
+            self._current_sample = 0
+            raise
+        return callback_out
+
+    def _feed_input(self, data, kwargs):
+        if self._callback is not None:
+            raise RuntimeError(f"Cannot use `feed_input` on the external source '{self._name}' with a `source`"
+                               " argument specified.")
+
+        self._feed_inputs.put((data, kwargs))
+
+    def _fetch(self, epoch_idx):
+        """Fetches data from callback or provided with feed_input."""
+
+        if self._callback is not None:
+            callback_out = self._get_batch(epoch_idx)
+            layout = self._layout
+            if self._num_outputs is not None:
+                for idx in range(self._num_outputs):
+                    if self._batch:
+                        raw_data = callback_out[idx]
```

Comment:
Is this correct? We just save last output as `raw_data` here?

Reply:
It was a mistake, fixed it.